### ISSUE TITLE
Best way to process many geocoding requests

### ISSUE THREAD

**Issue by Ismamai** [googlemaps/google-maps-services-java#176]
Not sure if there is mailing list or google group to make these questions.

What is the recommended way to process multiple geocoding requests ?

I need to process batches of request to be geocoded. What I do is create a GeoApiContext and I use java 8 streams to do multiple calls to newRequestMethod on GeocodingApi 
`GeocodingApi.newRequest(geoApiContext).address(newRequestString)`
and then I set the RequestCallback to process the answer.

What approach are you using for this scenario ?

**Comment by domesticmouse** [comment]
Assuming a stream is a thread of execution, and you have all the requests going through a single `geoApiContext` then you are using the API the way it is designed to be used. The request rate limit will be honoured and all that good stuff.

**Comment by Ismamai** [comment]
I have been measuring the Geocoding process and the average time per geocoding is 13.5 seconds with a max of 26 seconds. That seems to be too much.

**Comment by domesticmouse** [comment]
Leading to the fun question of how big is the batch of addresses to be geocoded?

**Comment by Ismamai** [comment]
I just tried with a 90 requests file. I will try with a bigger file later. Maybe 2500 to max out the free limits of my developer account :)

**Comment by domesticmouse** [comment]
Given you are on a free account, your rate limit is ten requests per second, so 90 requests should take 9 seconds. 2500 requests should take 250 seconds. This all seems to be working as intended to me.

**Comment by Ismamai** [comment]
Well I tried also with a paid account that has better QPS and results took also long time to process

These are some measurings (Units=ms)
With 

- development  key 10 qps 

    (LastValue=26075.0, Hits=90.0, Avg=13547.91111111111, Total=1219312.0, Min=834.0, Max=26113.0, First Access=Fri Aug 26 12:54:00 BST 2016, Last Access=Fri Aug 26 12:55:44 BST 2016) 

    **Total 104 seconds instead of theorical (90 request, 10 qps) 9 seconds**
- With paid key 50 qps

    (LastValue=22805.0, Hits=90.0, Avg=10768.577777777778, Total=969172.0, Min=522.0, Max=22805.0, First Access=Fri Aug 26 14:38:49 BST 2016, Last Access=Fri Aug 26 14:39:51 BST 2016)

    **Total 62 seconds instead of theorical (90 request, 50 qps) 1.8 seconds**

I know that my test is quite small and that I get network latency that is affecting the requests. 

This is why I am asking, just to be sure that I am not missing anything and I am getting maximum performance that I could get

**Comment by domesticmouse** [comment]
that doesn't sound right, can you put your test code up as a github repo?

**Comment by Ismamai** [comment]
I use batches of 30 requests then even that I could use 50 qps, I could only benefit of 30 qps using paid keys would be (90 requests, 30) ) 3 seconds. On any case the times I am getting are much bigger than expected.

I can not put the code I have used as it is now but I will try to isolate it. 

I saw the test in the library, but it only does one geocoding.

**Comment by Ismamai** [comment]
Sorry for the delay. I just pushed the code that allows to test the process to geocode addresses using the library and plain urls. https://github.com/Ismamai/performanceTest

With a Sample file of 90 addresses and using free keys:

- with google-maps-services-java library, it takes almost 14 seconds (theoretically should be around 9 seconds)

> real    0m13.869s
> user    0m0.015s
> sys     0m0.060s
- with plain url loading (this file can be build based on previous test or replacing key on provided file) takes aroung 7.4 seconds

> real    0m7.349s
> user    0m0.015s
> sys     0m0.060s

I was hoping the library to run much faster as it is loading the urls in paralled vs the plain url that is loading them sequentially.

I have tried also using paid keys, that have 50 QPS and should be much faster that free and plain url. It took almost 9 seconds (theoretically it should be around 3 seconds). 
> real    0m8.885s
> user    0m0.030s
> sys     0m0.061s

Not sure if I am missing something.

**Comment by Ismamai** [comment]
I used time command as it seems that JAMON library used to monitor request seems to give strange results.

**Comment by domesticmouse** [comment]
It is worth investigating the limit I built in the API. I may well have limited it to ten requests in flight as opposed to 10 requests initiated per second. Hmmm.

**Comment by Ismamai** [comment]
Using the last changes and the performanceTest 
I get this:

developent key and QPS=10
real 0m9.265s
user 0m0.000s
sys 0m0.075s

production key and QPS = 50
real 0m17.442s
user 0m0.000s
sys 0m0.105s
These time did not make any sense, then I rerun them and I got

real 0m8.500s
user 0m0.076s
sys 0m0.106s

And plain URL
real 0m7.492s
user 0m0.015s
sys 0m0.076s

**Comment by Ismamai** [comment]
I have tried setting the timeouts fields on GeoApiContext to 1 second and even less time but I do not see any change.
```
           geoApiContext.setConnectTimeout(1, TimeUnit.SECONDS)
           .setReadTimeout(1, TimeUnit.SECONDS)
           .setWriteTimeout(1, TimeUnit.SECONDS);

```

**Comment by domesticmouse** [comment]
It's going to take network traces to figure out where the time is being lost, I suspect.

**Comment by Ismamai** [comment]
I have found where the problem seems to be. In the RateLimitExecutorService it creates a ThreadPoolExecutor 
`  private final ExecutorService delegate = new ThreadPoolExecutor(0, Integer.MAX_VALUE, 60,
      TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(),
      threadFactory("Rate Limited Dispatcher", true));
`

The problem is that using corePoolSize set to 0 only creates a thread instead of creating many 
`  private final ExecutorService delegate = new ThreadPoolExecutor(50, Integer.MAX_VALUE, 60,
      TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(),
      threadFactory("Rate Limited Dispatcher", true));
`

With that change I got much better times.

Using the https://github.com/zvrablik/geocodingMockServer I was able to get almost 50 qps

Instead of ThreadPoolExecutor maybe it could use Executors.newWorkStealingPool(50) that grows and shrinks as needed.

**Comment by Ismamai** [comment]
I also tested rebuilding library with that change and using it with https://github.com/Ismamai/performanceTest and times were at least half.

**Comment by domesticmouse** [comment]
@Ismamai do you want to create a PR for this change?

**Comment by Ismamai** [comment]
Yes I will. I just created a new implementation of a geocoder client and
took the RateLimitExecutorService and with my tests I found that all my
calls were still sequential. Just changing the internal Executor fixed the
issue. In the library, the way this RateLimitExecutorService is used is
different to my usage, as you pass this class to OkHttpClient as
dispatcher. I will have to see how it fits best for the library.

**Comment by Ismamai** [comment]
I added to pullrequests with two different changes

[Using newWorkStealingPool](https://github.com/googlemaps/google-maps-services-java/pull/198)

[Using ThreadPoolExecutor](https://github.com/googlemaps/google-maps-services-java/pull/199)

I would be inclined to take approach with [newWorkStealingPool](https://github.com/googlemaps/google-maps-services-java/pull/198) as thread pool will grow and shrink as needed. But [ThreadPoolExecutor](https://github.com/googlemaps/google-maps-services-java/pull/199 seems to [](url)) option does not require to upgrade java version to 8 and allows to name threads with the ThreadFactory.

What do you think that is the best option?

**Comment by domesticmouse** [comment]
I'm going to call a friend for some tactical advice.

/cc @markmcd thoughts?

**Comment by domesticmouse** [comment]
I'm leaning towards using the ThreadPoolExecutor patch, if only so as not to change to much in this release. I'd be happier going to Java8 as part of upgrading to OkHttp3, with a more major version bump.

**Comment by Ismamai** [comment]
Well I just pushed the two options. I am not sure if the newWorkStealingPool would have any problem with OkHttpClient. But for sure it would be more efficient as it would grow and shrink when needed

**Comment by domesticmouse** [comment]
I've merged #199, so I believe we can close this bug about now. I'm going to close out #198 for the time being. We can revisit after I've uplifted to OkHttp3. Thanks!
